### PR TITLE
Switch from the deprecated imp to importlib

### PR DESCRIPTION
**Pull request by forslund**
## Description
Replaces the current imp based skill loader with an importlib based solution. This also adds a small change allowing better use of submodules. Before submodules were never reloaded forcing the over-use of the `__init__.py` file in skills. The change makes the submodules reload together with the main module by removing the refs from `sys.modules`, allowing better structure of skills.

## How to test
Make sure skills load. Make a change to a skill and ensure that the skill is reloaded and that the change is accessible.
Add a log statement to the Spotify skill's spotify.py and check that it's included when the skill reloads.

## Contributor license agreement signed?
CLA [ Yes ]

### CODE REVIEW

**Comment by chrisveilleux** [comment]
If we are going to do this, I want it to include the ability to load the code from a file other than `__init__.py`.  Putting all your code in this file is a Python anti-pattern.  This PR should make it possible for an author to move their skill code from `__init__.py` to something like `skill.py`.  At the very least, it should make it possible for __init__.py to be able to import code from another file, like `skill.py` and then contain only the import and the `create_skill` function.

**Comment by forslund** [comment]
Isn't that already possible?

The only thing that's needed is that `__init__.py` has the `create_skill()` function, either locally or imported from a module...

**Comment by chrisveilleux** [comment]
I moving the skill class to a file called `skill.py` a while ago and importing that into `__init__.py`.  it did not work.  It is possible that I did it wrong.  It is also possible that the new version of importlib supports this whereas the older imp library did not.  

I can test against the code in this PR and see if there is different behavior.

**Comment by forslund** [comment]
Ah, right imported submodules might not be reloaded...let me do some tests here

**Comment by forslund** [comment]
Well the [spotify skill](https://github.com/forslund/spotify-skill/blob/19.08/__init__.py#L38) has the majority of the spotify logic in a separate file so it has worked before at least...

let me try with your case with the entire skill in a skill.py file...

**Comment by chrisveilleux** [comment]
I would be happy if we just didn't hardcode the `__init__.py` as the only place skill code can come from.  Make it so that the code can be loaded from there or from another location, like `skill.py`.  We can support both for a time then remove support for `__inti__.py`.

**Comment by forslund** [comment]
Do you mean like search all files for `create_skill()`?

Looking at a skill as a module it would require the `__init__.py` to show what's exported right?

**Comment by chrisveilleux** [comment]
using `__init__.py` as a way to expose the API of a package is not a Python requirement.  Lots of folks just leave that file completely empty.  It is my preference the `__init__.py` be used in all of our code in this way.  So if we can get that to work with importlib, that would be ideal.  Ideally, we could import `create_skill()` from skill.py, load it in the skill loader and execute it.  I just don't know off the top of my head what is possible and what is not with importlib.  If we need to do something a bit different to support how we load skills the way we do, I am OK with that.

**Comment by chrisveilleux** [comment]
SHIT.  Hit the close button accidentally.

**Comment by forslund** [comment]
Gotcha.

In any case you're query highlighted an issue here. with this implementation relative imports no longer works in skills so I guess I need to look into that at least.

Will look at loading the skill from any file while I'm at it.

**Comment by JarbasAl** [comment]
> Gotcha.
> 
> In any case you're query highlighted an issue here. with this implementation relative imports no longer works in skills so I guess I need to look into that at least.
> 
> Will look at loading the skill from any file while I'm at it.

relative imports haven't worked ever, i remember in the old times the usual suggestion in slack was to add the skill directory to the python path, i usually started from MYCROFT_ROOT_PATH but i remember seeing all sorts of ugly stuff with os.path and some approaches with skill variables

**Comment by chrisveilleux** [comment]
In case there was any question, manipulating the Python path should be avoided at all costs.

**Comment by ChanceNCounter** [comment]
> relative imports haven't worked ever, i remember in the old times the usual suggestion in slack was to add the skill directory to the python path, i usually started from MYCROFT_ROOT_PATH (remember when skill where part of core?) but i remember seeing all sorts of ugly stuff with os.path and some approaches with skill variables

I'm using relative imports in a skill I'm developing right now, and it seems to be working. Aside from this PR, is that going to break later? There seem to be conflicting experiences here.

**Comment by forslund** [comment]
With the current `imp` implentation it works. I did some work last night and managed to get the relative imports working with importlib as well. It would be terrible if the skills were limited to a single file.

There is an issue with the relative imports not reloading when the `__init__.py` is reloaded (both in old imp and importlib implementations) but I think I've got a handle on how to tackle that issue.

**Comment by forslund** [comment]
Pushed a WIP update supporting relative imports again and adding reloading for them as well...

**Comment by forslund** [comment]
This has been updated to use importlib and a nice and easy fix to allow skill submodules to be reloaded.

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2327))

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2327))

File: mycroft/skills/skill_loader.py [hunk 1]
```diff
@@ -29,6 +34,49 @@
 SKILL_MAIN_MODULE = '__init__.py'
 
 
+def remove_submodule_refs(module_name):
+    """Ensure submodules being reloaded by removing the refs from sys.modules.
```
Comment (krisgesling) [inline mycroft/skills/skill_loader.py:38]: #nitpick...

> Ensure submodule is reloaded by removing the refs from sys.modules.

File: mycroft/skills/skill_loader.py [hunk 2]
```diff
@@ -29,6 +34,49 @@
...
+
+    Python import system puts a reference for each module in the sys.modules
+    dictionary to bypass loading if a module is already in memory. To make
+    sure skills are completely reloaded these references are deleted.
+
+    Arguments:
+        module_name: name of skill module.
+    """
+    submodules = []
+    LOG.info('Skill module'.format(module_name))
```
Comment (krisgesling) [inline mycroft/skills/skill_loader.py:48]: is this meant to be info level?

File: mycroft/skills/skill_loader.py (hunk 1 above)
Comment (krisgesling) [inline mycroft/skills/skill_loader.py:38]: #nitpick...

> Ensure submodule is reloaded by removing the refs from sys.modules.

File: mycroft/skills/skill_loader.py (hunk 2 above)
Comment (krisgesling) [inline mycroft/skills/skill_loader.py:48]: Is this intended to be an info level log message?

Comment (forslund) [reply mycroft/skills/skill_loader.py:48]: You're right that should be a debug level, will fix on my train ride home

File: mycroft/skills/skill_loader.py (hunk 1 above)
Comment (forslund) [reply mycroft/skills/skill_loader.py:38]: Are you ok with

>Ensure submodules are reloaded by removing the refs from sys.modules.

Comment (krisgesling) [reply mycroft/skills/skill_loader.py:38]: Yep that reads great.

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2327))

File: mycroft/skills/skill_loader.py (hunk 1 above)
Comment (forslund) [reply mycroft/skills/skill_loader.py]: Comment updated

File: mycroft/skills/skill_loader.py (hunk 2 above)
Comment (forslund) [reply mycroft/skills/skill_loader.py]: Changed to debug

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2327))

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2327))

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2327))

**Pull request merged**